### PR TITLE
fix(mssql): reverts #9644 and displays a better error msg

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

- [X] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
On SQLLAb executing SQL statements with `CAST` and `AT` fail. This is a partial fix, since this seems to open up a bigger problem `sqlparse` does dot correctly identify columns with `CAST(FOO_COL AS TYPEX) AT TIME ZONE 'Eastern Standard Time'` for example, and just adds `CAST(FOO_OLD AS TYPEX)` on the IdentifierList.

This is a partial fix, since it will discard `CAST` from the alias setting, so that we don't double alias it. Yet it means that all `CAST` statements need to have an Alias.

I think we should consider removing our forced `TOP` from MSSQL engines and maybe others that use `TOP`, too much complexity for the value it adds.

PS: Following comments I'm reverting all logic behind squeezing aliases on functions for MSSQL. This implements a much simpler approach by catching MSSQL specific error for non aliased functions. This error is easy to catch since we are wrapping a TOP on the user's statement, otherwise we get an error from superset saying `no field of name`

Examples:
Screenshot 2020-05-12 at 13 08 02

Screenshot 2020-05-12 at 13 26 58

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
This is really a can of worms, and the deeper we go the more `sqlparse`-like logic we'll be introducing into Superset with unpleasant maintenance burden. I would almost propose deprecating advanced limiting for `TOP`-syntax dialects.

**Comment by mistercrunch** [comment]
The problem is not as much `TOP` as it is the requirement for alias in subquery.

And oh wow! I didn't know we were altering the SQL (as opposed to just wrapping it). That's pretty scary. Whenever parsing or alerting SQL is part of the solution, it's pretty scary. I'd advocate for even rolling back the logic that's in there currently (from before this PR).

Also if this logic should be anywhere, it should be scoped to MSSQL in `db_engine_spec` module

Few other options:
1. force a TOP clause (similar to `LimitMethod.FORCE_LIMIT`) it is altering the SQL, but somewhat less risky than squeezing aliases in. Some optimizers will do better with that than with the `LimitMethod.WRAP_SQL` approach.
1. surface a good error message to user "please alias all of your columns" and move forward with `LimitMethod.WRAP_SQL`
1. use `limit_method = LimitMethod.FETCH_MANY`, this has bad perf implications (usually means a larger "server side cursor" than necessary)
1. long shot: look for a way to let the server know ahead of time, some sort of cursor hint or parameter. That's clearly out of the DBAPI specification but it may exist

Personally I think 1 is best. It has the best perf guarantees (gives the clearest declarative insight to the db optimizer) and requires limited query alteration. When we moved forward with this a while ago I remember we looked at how other tools did this but forgot the details, maybe @bkyryliuk remembers?

**Comment by dpgaspar** [comment]
thks for the ideas @mistercrunch, we had/have a couple of problems with MSSQL on SQLLab. 
 1 - TOP wrapping was not working properly
 2 - Even without TOP wrapping, functions on MSSQL like `COUNT`, `SUM` etc are set has `<anonymous> and these are not supported by superset (pandas dataframes?)

Given that, maybe 2 is a good approach, or a mix of 1 and 2. May be wrong here, but still think that altering SQL statements is dangerous/difficult and kind of out of scope for superset. Another option could be not forcing `TOP` on some engines but refusing to execute statements that don't have them? 

Note: This is scoped to MSSQL, because it was an obvious risky change

**Comment by bkyryliuk** [comment]
yeah we've looked into alternatives quite some time ago and sqlparse was the only feasible option at a time. There are DB specific parsers, but they often implemented in different languages e.g. java + antlr for hive and presto as I remember and using them as was mentioned will be opening a can of warms. Explain query can provide some useful information as well, however I am not sure about MS SQL specifics there. I'm curious what is a sqllachemy solution if it supports query construction for MSSQL

**Comment by dpgaspar** [comment]
@bkyryliuk we have a simple query construction here: https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/base.py#L397 and it's being used for MSSQL. Is something similar what you're looking for?

@mistercrunch also mentioned that using a `WRAP_SQL` is inefficient on some engines.

My plan is, if feasible/simple change `WRAP_SQL` to something like `FORCE_TOP` and doing the best to identify non aliased functions and warn the user about it. This last one could be tricky for statements like: `SELECT COL1, COL1, CAST(COL3 AS DATETIME) AT TIME ZONE "SOME_TZ", COL4` `sqlparse` get seems to get lost on `AT` and does not identify `COL4` has a column.

**Comment by bkyryliuk** [comment]
I like the approach, and eventually it would be nice to have that logic in the db specific engines.
I briefly looked into explain and it doesn't look promising, https://docs.microsoft.com/en-us/sql/t-sql/queries/explain-transact-sql?view=azure-sqldw-latest

Another potential alternative could be to catch an error and make it more user friendly e.g. stating that all columns should be aliased. It would make user experience slightly worse in the beginning, but over time should not be an issue and may be slightly easier to implement.

**Comment by mistercrunch** [comment]
For reference, here's the FORCE_LIMIT logic:
https://github.com/apache/incubator-superset/blob/13c5b133a9d054b07dff5bbcff987551e3c4c42e/superset/sql_parse.py#L255-L286

FORCE_TOP wouldn't be that different. It's editing the SQL, but it's pretty safe (not as bad as squeezing aliases into subqueries...)

**Comment by dpgaspar** [comment]
Yes I know and totally agree @mistercrunch bad initial approach on my side.

@bkyryliuk catching the error may be a good approach, could turn out to be difficult to pin the engine error, specific to non aliased functions, on the other hand, it's also a bit hard to parse the query and search for non aliased functions. So I may give it a try

**Comment by dpgaspar** [comment]
Updated this PR and description with:
```
Following comments I'm reverting all logic behind squeezing aliases on functions for MSSQL. This implements a much simpler approach by catching MSSQL specific error for non aliased functions. This error is easy to catch since we are wrapping a TOP on the user's statement, otherwise we get an error from superset saying no field of name
```

File: superset/db_engine_specs/mssql.py [hunk 1]
```diff
@@ -85,6 +84,10 @@ def get_sqla_column_type(cls, type_: str) -> Optional[TypeEngine]:
...
 
     @classmethod
-    def apply_limit_to_sql(cls, sql: str, limit: int, database: "Database") -> str:
-        new_sql = ParsedQuery(sql).set_alias()
-        return super().apply_limit_to_sql(new_sql, limit, database)
+    def extract_error_message(cls, ex: Exception) -> str:
+        if str(ex).startswith("(8155"):
+            return (
+                f"{cls.engine} error: All your SQL functions need to "
+                "have alias on MSSQL. For example: SELECT COUNT(*) AS C1 FROM TABLE1"
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py:91]:
```suggestion
                "have an alias on MSSQL. For example: SELECT COUNT(*) AS C1 FROM TABLE1"
```

File: superset/db_engine_specs/mssql.py [hunk 2]
```diff
@@ -85,6 +84,10 @@ def get_sqla_column_type(cls, type_: str) -> Optional[TypeEngine]:
         return None
 
     @classmethod
-    def apply_limit_to_sql(cls, sql: str, limit: int, database: "Database") -> str:
-        new_sql = ParsedQuery(sql).set_alias()
-        return super().apply_limit_to_sql(new_sql, limit, database)
+    def extract_error_message(cls, ex: Exception) -> str:
+        if str(ex).startswith("(8155"):
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py:88]:
```suggestion
        if str(ex).startswith("(8155,"):
```
You never know if they introduce `^\(8155\d+,`..

File: tests/db_engine_specs/mssql_tests.py [hunk 3]
```diff
@@ -97,64 +94,28 @@ def test_convert_dttm(self):
...
-            expected_sql = (
-                "SELECT TOP 1000 * \n"
-                "FROM (SELECT COUNT(*) AS COUNT_1, COUNT(*) AS COUNT_2 FROM FOO_TABLE)"
-                " AS inner_qry"
-            )
-            self.assertEqual(expected_sql, limited_sql)
+        error_message = MssqlEngineSpec.extract_error_message(test_mssql_exception)
+        expected_message = (
+            "mssql error: All your SQL functions need to "
+            "have alias on MSSQL. For example: SELECT COUNT(*) AS C1 FROM TABLE1"
```
Comment (villebro) [inline tests/db_engine_specs/mssql_tests.py:106]:
```suggestion
            "have an alias on MSSQL. For example: SELECT COUNT(*) AS C1 FROM TABLE1"
```

**Review by villebro — CHANGES_REQUESTED** [review]
A few minor suggestions

File: superset/db_engine_specs/mssql.py (hunk 2 above)
Comment (dpgaspar) [reply superset/db_engine_specs/mssql.py:88]: good point

**Review by villebro — APPROVED** [review]
